### ISSUE TITLE
Support GraphQL Default Variables

### ISSUE THREAD

**Issue by ethansinjin** [apollographql/apollo-ios#212]
According to the [GraphQL documentation](http://graphql.org/learn/queries/#default-variables), default variables can be specified like this:

```
query HeroNameAndFriends($episode: Episode = "JEDI") {
  hero(episode: $episode) {
    name
    friends {
      name
    }
  }
}
```

However, currently Apollo iOS generates something like the following initializer for a scenario like this:

```
public init(episode: Episode? = nil) {
  self.episode = episode
}
```

It would be ideal if Apollo were able to do something like this:

```
public init(episode: Episode? = Episode(string: "Jedi") {

}
```

At the very least, there should be a warning that this behavior is not occurring, either in the documentation or during compile time. Our team has run into subtle bugs due to this discrepancy in behavior.

**Comment by psi-gh** [comment]
Have issue with empty default variables. In my query I have variable 
    query Fetch($id:ID!, $withFriends:Boolean = true) 
But in json query which goes to server I have:
    "variables":{"id":"1cdfb191-8fe0-4064-b452-bca90dc35bd9","withFriends":null}
instead of 
    "variables":{"id":"1cdfb191-8fe0-4064-b452-bca90dc35bd9","withFriends":true}

It leads to incorrect answer.

**Comment by koenpunt** [comment]
This is also an issue when using default values for `@include(if: ...)` statements, since those require the value to be set.

Take for example this query;

```graphql
query RelatedProducts(
  $productId: ID!
  $withDiscount: Boolean = true
) {
  product(id: $productId) {
    relatedProductsConnection(first: $first) {
      nodes {
        id
        discount @include(if: $withDiscount) {
          ...DiscountFragment
        }
      }
    }
  }
}
```

The generated struct initializer looks good;

```swift
public final struct RelatedProductsQuery: GraphQLQuery {
  // ...
  public init(productId: GraphQLID, withDiscount: Bool? = nil) {
    self.productId = productId
    self.withDiscount = withDiscount
  }
}
```

But when serialized, the arguments look like this:

```json
{
  "productId": "123",
  "withDiscount": null
}
```

Which causes the `@include(if: ...)` to bark;

```
Thread 1: Fatal error: Error: Argument "if" of non-null type "Boolean!" must not be null.
```

But if the serialized arguments would look like this:

```json
{
  "productId": "123"
}
```

The request would work just fine.

So preferably parameters that are defined with a default value shouldn't be included in the serialized arguments if no value is set for them.

**Comment by designatednerd** [comment]
Yep - this is something I'll be trying to take a look at as part of the Swift Codegen project,

**Comment by kaushik143** [comment]
@koenpunt Did you find any workaround for your issue? I am facing a similar issue when I migrating to apollo-v2 and now since there are ios client on the older version I need some workaround on graphql to make it work and ignore this check

**Comment by koenpunt** [comment]
@kaushik143 it's been a while, and I'm no longer working on that project, so I'm not sure, but I think we just made just that we always passed an argument.

**Comment by kaushik143** [comment]
@koenpunt No worries. We are also planning for a similar change where we pass the default value always my only worry is we will have to release force updates after these changes. @designatednerd Do you have any workaround for this on the graphql server? May be overriding include directive?

**Comment by designatednerd** [comment]
I think there's some variation per server on what happens. On our Star Wars Server if an `Episode` is not provided when searching for a `hero`, the server uses the default value on the server (in this case, `NEWHOPE`). This means that if the default value is changed after the code is generated, it doesn't matter because the server will always know what the default is, and will use it unless a value is explicitly provided.

But some servers have things like this as a non-nullable property, in which case passing whatever the default value is at the time you're generating the code is going to be your best bet.

**Comment by calvincestari** [comment]
Hi 👋🏻 - this issue should be resolved in 1.0. 1.0 is currently in alpha releases with a full 1.0 release in H2 2022. Please take a look at the [latest alpha](https://github.com/apollographql/apollo-ios/releases) and let us know if this is still an issue for you.
